src/v/config/property.h

```diff
@@ -523,4 +523,75 @@ class deprecated_property : public property<ss::sstring> {
 
     void set_value(std::any) override { return; }
 };
+
+template<typename T>
+class enum_property : public property<T> {
+public:
+    enum_property(
+      config_store& conf,
+      std::string_view name,
+      std::string_view desc,
+      base_property::metadata meta,
+      T def,
+      std::vector<T> values)
+      : property<T>(
+        conf,
+        name,
+        desc,
+        meta,
+        def,
+        [this](T new_value) -> std::optional<ss::sstring> {
+            auto found = std::find_if(
+              _values.begin(), _values.end(), [&new_value](T const& v) {
+                  return v == new_value;
+              });
+            if (found == _values.end()) {
+                return help_text();
+
+            } else {
+                return std::nullopt;
+            }
+        })
+      , _values(values) {}
+
+    std::optional<validation_error>
+    validate(YAML::Node n) const final override {
+        try {
+            auto v = n.as<T>();
+            return property<T>::validate(v);
+        } catch (...) {
+            // Not convertible (e.g. if the underlying type is an enum class)
+            // therefore assume it is out of bounds.
+            return validation_error{property<T>::name().data(), help_text()};
+        }
+    }
+
+    std::vector<ss::sstring> enum_values() const final override {
+        std::vector<ss::sstring> r;
+        for (const auto& v : _values) {
+            r.push_back(ssx::sformat("{}", v));
+        }
+
+        return r;
+    }
+
+private:
+    ss::sstring value_list() const {
+        std::ostringstream str;
+        for (size_t i = 0; i < _values.size(); ++i) {
+            str << _values[i];
+            if (i != _values.size() - 1) {
+                str << ", ";
+            }
+        }
+        return str.str();
```

Comment:
nit: `fmt::join`

Reply:
Changed to use join, although not entirely slick because the underlying type isn't necessarily formattable, so there's still a preamble where I build a list of string-ized values.

Reply:
oh that is really unsatisfying haha